xs/src/libslic3r/MedialAxis.cpp

```diff
@@ -0,0 +1,1477 @@
+            if ((polyline.endpoints.first || polyline.endpoints.second)
+                && polyline.length() < max_width / 2) {
+                if (shortest_size > polyline.length()) {
+                    shortest_size = polyline.length();
+                    shortest_idx = i;
+                }
+
+            }
+        }
+        if (shortest_idx >= 0 && shortest_idx < pp.size()) {
+            pp.erase(pp.begin() + shortest_idx);
+            changes = true;
+        }
+        if (changes) concatThickPolylines(pp);
+    }
+}
+
+void
+MedialAxis::ensure_not_overextrude(ThickPolylines& pp)
+{
+    //ensure the volume extruded is correct for what we have been asked
+    // => don't over-extrude
+    double surface = 0;
+    double volume = 0;
+    for (ThickPolyline& polyline : pp) {
+        for (ThickLine l : polyline.thicklines()) {
+            surface += l.length() * (l.a_width + l.b_width) / 2;
+            double width_mean = (l.a_width + l.b_width) / 2;
+            volume += height * (width_mean - height * (1. - 0.25 * PI)) * l.length();
+        }
+    }
+
+    // compute bounds volume
+    double boundsVolume = 0;
+    boundsVolume += height*bounds.area();
+    // add external "perimeter gap"
+    double perimeterRoundGap = bounds.contour.length() * height * (1 - 0.25*PI) * 0.5;
+    // add holes "perimeter gaps"
+    double holesGaps = 0;
+    for (auto hole = bounds.holes.begin(); hole != bounds.holes.end(); ++hole) {
+        holesGaps += hole->length() * height * (1 - 0.25*PI) * 0.5;
+    }
+    boundsVolume += perimeterRoundGap + holesGaps;
+
+    if (boundsVolume < volume) {
+        //reduce width
+        double reduce_by = boundsVolume / volume;
+        for (ThickPolyline& polyline : pp) {
+            for (ThickLine l : polyline.thicklines()) {
+                l.a_width *= reduce_by;
+                l.b_width *= reduce_by;
+            }
+        }
+    }
+}
+
+ExPolygon 
+MedialAxis::simplify_polygon_frontier()
+{
+
+    //simplify the boundary between us and the bounds.
+    //int firstIdx = 0;
+    //while (firstIdx < contour.points.size() && bounds.contour.contains(contour.points[firstIdx])) firstIdx++;
+    ExPolygon simplified_poly = this->surface;
+    if (&this->surface != &this->bounds) {
+        bool need_intersect = false;
+        for (size_t i = 0; i < simplified_poly.contour.points.size(); i++) {
+            Point &p_check = simplified_poly.contour.points[i];
+            //if (!find) {
+            if (!bounds.has_boundary_point(p_check)) {
+                //check if we put it at a bound point instead of delete it
+                size_t prev_i = i == 0 ? simplified_poly.contour.points.size() - 1 : (i - 1);
+                size_t next_i = i == simplified_poly.contour.points.size() - 1 ? 0 : (i + 1);
+                const Point* closest = bounds.contour.closest_point(p_check);
+                if (closest != nullptr && closest->distance_to(p_check) + SCALED_EPSILON
```

Comment:
This can move point quite far ... is that desirable?

Reply:
It's on the frontier/interface/polyline with the perimeter. It's a small area, we will extrude into it, and it can create artifacts if it has smalls zigs. Simplifying it aggressively has improved greatly the output of the medial_axis.